File: packages/fmlrunner/src/lib/bundle-service.ts
import { Bundle, BundleEntry, StructureMap, StructureDefinition, ConceptMap, ValueSet, CodeSystem } from '../types';
import { ConceptMapService } from './conceptmap-service';
import { ValueSetService } from './valueset-service';
import { CodeSystemService } from './codesystem-service';
import { ValidationService } from './validation-service';
import { Logger } from './logger';

/**
 * Result of processing a bundle
 */
export interface BundleProcessingResult {
  success: boolean;
  processed: {
    structureMaps: number;
    structureDefinitions: number;
    conceptMaps: number;
    valueSets: number;
    codeSystems: number;
    other: number;
  };
  errors: string[];
  warnings: string[];
}

/**
 * Service for processing FHIR Bundles and distributing resources to appropriate services
 */
export class BundleService {
  private logger: Logger;
  
  constructor(
    private conceptMapService: ConceptMapService,
    private valueSetService: ValueSetService,
    private codeSystemService: CodeSystemService,
    private validationService: ValidationService | undefined,
    private structureMapStore: Map<string, StructureMap>,
    logger: Logger
  ) {
    this.logger = logger;
  }

  /**
   * Process a FHIR Bundle and register all contained resources
   */
  processBundle(bundle: Bundle): BundleProcessingResult {
    const result: BundleProcessingResult = {
      success: true,
      processed: {
        structureMaps: 0,
        structureDefinitions: 0,
        conceptMaps: 0,
        valueSets: 0,
        codeSystems: 0,
        other: 0
      },
      errors: [],
      warnings: []
    };

    if (!bundle.entry || bundle.entry.length === 0) {
      result.warnings.push('Bundle contains no entries');
      return result;
    }

    for (let i = 0; i < bundle.entry.length; i++) {
      const entry = bundle.entry[i];
      
      try {
        this.processEntry(entry, i, result);
      } catch (error) {
        const errorMsg = `Error processing entry ${i}: ${error instanceof Error ? error.message : 'Unknown error'}`;
        result.errors.push(errorMsg);
        result.success = false;
      }
    }

    return result;
  }

  /**
   * Process a single bundle entry
   */
  private processEntry(entry: BundleEntry, index: number, result: BundleProcessingResult): void {
    if (!entry.resource) {
      result.warnings.push(`Entry ${index} has no resource`);
      return;
    }

    const resource = entry.resource;
    
    switch (resource.resourceType) {
      case 'StructureMap':
        this.processStructureMap(resource as StructureMap, index, result);
        break;
        
      case 'StructureDefinition':
        this.processStructureDefinition(resource as StructureDefinition, index, result);
        break;
        
      case 'ConceptMap':
        this.processConceptMap(resource as ConceptMap, index, result);
        break;
        
      case 'ValueSet':
        this.processValueSet(resource as ValueSet, index, result);
        break;
        
      case 'CodeSystem':
        this.processCodeSystem(resource as CodeSystem, index, result);
        break;
        
      default:
        result.processed.other++;
        result.warnings.push(`Entry ${index}: Unsupported resource type '${resource.resourceType}'`);
    }
  }

  /**
   * Process StructureMap resource
   */
  private processStructureMap(structureMap: StructureMap, index: number, result: BundleProcessingResult): void {
    try {
      if (!structureMap.id && !structureMap.url) {
        result.warnings.push(`Entry ${index}: StructureMap has no id or url, skipping`);
        return;
      }

      // Store in StructureMap store if available
      if (this.structureMapStore) {
        if (structureMap.id) {
          this.structureMapStore.set(structureMap.id, structureMap);
        }
        if (structureMap.url) {
          this.structureMapStore.set(structureMap.url, structureMap);
        }
      }

      result.processed.structureMaps++;
    } catch (error) {
      result.errors.push(`Entry ${index}: Failed to process StructureMap - ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Process StructureDefinition resource
   */
  private processStructureDefinition(structureDefinition: StructureDefinition, index: number, result: BundleProcessingResult): void {
    try {
      if (!structureDefinition.id && !structureDefinition.url) {
        result.warnings.push(`Entry ${index}: StructureDefinition has no id or url, skipping`);
        return;
      }

      // Register with validation service if available
      if (this.validationService) {
        this.validationService.registerStructureDefinition(structureDefinition);
      }

      result.processed.structureDefinitions++;
    } catch (error) {
      result.errors.push(`Entry ${index}: Failed to process StructureDefinition - ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Process ConceptMap resource
   */
  private processConceptMap(conceptMap: ConceptMap, index: number, result: BundleProcessingResult): void {
    try {
      if (!conceptMap.id && !conceptMap.url) {
        result.warnings.push(`Entry ${index}: ConceptMap has no id or url, skipping`);
        return;
      }

      this.conceptMapService.registerConceptMap(conceptMap);
      result.processed.conceptMaps++;
    } catch (error) {
      result.errors.push(`Entry ${index}: Failed to process ConceptMap - ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Process ValueSet resource
   */
  private processValueSet(valueSet: ValueSet, index: number, result: BundleProcessingResult): void {
    try {
      if (!valueSet.id && !valueSet.url) {
        result.warnings.push(`Entry ${index}: ValueSet has no id or url, skipping`);
        return;
      }

      this.valueSetService.registerValueSet(valueSet);
      result.processed.valueSets++;
    } catch (error) {
      result.errors.push(`Entry ${index}: Failed to process ValueSet - ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Process CodeSystem resource
   */
  private processCodeSystem(codeSystem: CodeSystem, index: number, result: BundleProcessingResult): void {
    try {
      if (!codeSystem.id && !codeSystem.url) {
        result.warnings.push(`Entry ${index}: CodeSystem has no id or url, skipping`);
        return;
      }

      this.codeSystemService.registerCodeSystem(codeSystem);
      result.processed.codeSystems++;
    } catch (error) {
      result.errors.push(`Entry ${index}: Failed to process CodeSystem - ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Create a summary bundle of all loaded resources
   */
  createSummaryBundle(): Bundle {
    const entries: BundleEntry[] = [];
    
    // Add StructureMaps
    if (this.structureMapStore) {
      const uniqueStructureMaps = new Map<string, StructureMap>();
      this.structureMapStore.forEach((sm) => {
        const key = sm.id || sm.url || Math.random().toString();
        uniqueStructureMaps.set(key, sm);
      });
      
      uniqueStructureMaps.forEach((sm) => {
        entries.push({
          fullUrl: sm.url || `StructureMap/${sm.id}`,
          resource: sm
        });
      });
    }

    // Add StructureDefinitions
    if (this.validationService) {
      const structureDefinitions = this.validationService.getStructureDefinitions();
      structureDefinitions.forEach((sd) => {
        entries.push({
          fullUrl: sd.url || `StructureDefinition/${sd.id}`,
          resource: sd
        });
      });
    }

    // Add ConceptMaps
    this.conceptMapService.getAllConceptMaps().forEach((cm) => {
      entries.push({
        fullUrl: cm.url || `ConceptMap/${cm.id}`,
        resource: cm
      });
    });

    // Add ValueSets
    this.valueSetService.getAllValueSets().forEach((vs) => {
      entries.push({
        fullUrl: vs.url || `ValueSet/${vs.id}`,
        resource: vs
      });
    });

    // Add CodeSystems
    this.codeSystemService.getAllCodeSystems().forEach((cs) => {
      entries.push({
        fullUrl: cs.url || `CodeSystem/${cs.id}`,
        resource: cs
      });
    });

    return {
      resourceType: 'Bundle',
      id: 'loaded-resources-' + Date.now(),
      type: 'collection',
      timestamp: new Date().toISOString(),
      total: entries.length,
      entry: entries
    };
  }

  /**
   * Clear all loaded resources
   */
  clearAll(): void {
    this.conceptMapService.clear();
    this.valueSetService.clear();
    this.codeSystemService.clear();
    if (this.structureMapStore) {
      this.structureMapStore.clear();
    }
  }

  /**
   * Get loading statistics
   */
  getStats(): {
    structureMaps: number;
    structureDefinitions: number;
    conceptMaps: number;
    valueSets: number;
    codeSystems: number;
  } {
    return {
      structureMaps: this.structureMapStore ? Array.from(new Set(Array.from(this.structureMapStore.values()).map(sm => sm.id || sm.url))).length : 0,
      structureDefinitions: this.validationService ? this.validationService.getStructureDefinitions().length : 0,
      conceptMaps: this.conceptMapService.getCount(),
      valueSets: this.valueSetService.getCount(),
      codeSystems: this.codeSystemService.getCount()
    };
  }
}

File: src/commonMain/kotlin/org/litlfred/fmlrunner/terminology/BundleService.kt
package org.litlfred.fmlrunner.terminology

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*

/**
 * FHIR Bundle resource definition for batch processing
 * Based on FHIR R4 Bundle specification
 */
@Serializable
data class Bundle(
    val resourceType: String = "Bundle",
    val id: String? = null,
    val identifier: Identifier? = null,
    val type: String, // document | message | transaction | transaction-response | batch | batch-response | history | searchset | collection
    val timestamp: String? = null,
    val total: Int? = null,
    val link: List<BundleLink>? = null,
    val entry: List<BundleEntry>? = null,
    val signature: String? = null // Digital signature
)

@Serializable
data class BundleLink(
    val relation: String,
    val url: String
)

@Serializable
data class BundleEntry(
    val link: List<BundleLink>? = null,
    val fullUrl: String? = null,
    val resource: JsonElement? = null,
    val search: BundleEntrySearch? = null,
    val request: BundleEntryRequest? = null,
    val response: BundleEntryResponse? = null
)

@Serializable
data class BundleEntrySearch(
    val mode: String? = null, // match | include | outcome
    val score: Double? = null
)

@Serializable
data class BundleEntryRequest(
    val method: String, // GET | HEAD | POST | PUT | DELETE | PATCH
    val url: String,
    val ifNoneMatch: String? = null,
    val ifModifiedSince: String? = null,
    val ifMatch: String? = null,
    val ifNoneExist: String? = null
)

@Serializable
data class BundleEntryResponse(
    val status: String,
    val location: String? = null,
    val etag: String? = null,
    val lastModified: String? = null,
    val outcome: JsonElement? = null
)

/**
 * Result of Bundle processing operation
 */
@Serializable
data class BundleProcessingResult(
    val success: Boolean,
    val processed: ProcessedCounts,
    val errors: List<String> = emptyList(),
    val warnings: List<String> = emptyList()
)

@Serializable
data class ProcessedCounts(
    val structureMaps: Int = 0,
    val structureDefinitions: Int = 0,
    val conceptMaps: Int = 0,
    val valueSets: Int = 0,
    val codeSystems: Int = 0,
    val other: Int = 0
)

/**
 * Bundle processing statistics
 */
@Serializable
data class BundleStats(
    val totalBundles: Int = 0,
    val totalResources: Int = 0,
    val resourceCounts: Map<String, Int> = emptyMap(),
    val lastProcessed: String? = null
)

/**
 * Bundle Service using kotlin-fhir compatible types
 * This replaces the TypeScript BundleService with a kotlin-fhir based implementation
 */
class BundleService(
    private val conceptMapService: ConceptMapService,
    private val valueSetService: ValueSetService,
    private val codeSystemService: CodeSystemService,
    private val validationService: ValidationService
) {
    private var stats = BundleStats()
    private val resourceCounts = mutableMapOf<String, Int>()

    /**
     * Process a Bundle and distribute resources to appropriate services
     */
    fun processBundle(bundle: Bundle): BundleProcessingResult {
        val errors = mutableListOf<String>()
        val warnings = mutableListOf<String>()
        val processed = ProcessedCounts()

        try {
            val entries = bundle.entry ?: emptyList()
            var processedCounts = ProcessedCounts()

            for (entry in entries) {
                val resource = entry.resource
                if (resource != null) {
                    val result = processResource(resource)
                    processedCounts = processedCounts.copy(
                        structureMaps = processedCounts.structureMaps + result.structureMaps,
                        structureDefinitions = processedCounts.structureDefinitions + result.structureDefinitions,
                        conceptMaps = processedCounts.conceptMaps + result.conceptMaps,
                        valueSets = processedCounts.valueSets + result.valueSets,
                        codeSystems = processedCounts.codeSystems + result.codeSystems,
                        other = processedCounts.other + result.other
                    )
                }
            }

            // Update statistics
            updateStats(bundle, processedCounts)

            return BundleProcessingResult(
                success = errors.isEmpty(),
                processed = processedCounts,
                errors = errors,
                warnings = warnings
            )

        } catch (e: Exception) {
            errors.add("Bundle processing failed: ${e.message}")
            return BundleProcessingResult(
                success = false,
                processed = ProcessedCounts(),
                errors = errors,
                warnings = warnings
            )
        }
    }

    /**
     * Process a single resource and route to appropriate service
     */
    private fun processResource(resource: JsonElement): ProcessedCounts {
        if (resource !is JsonObject) {
            return ProcessedCounts(other = 1)
        }

        val resourceType = resource["resourceType"]?.jsonPrimitive?.content
        
        return when (resourceType) {
            "ConceptMap" -> {
                try {
                    val conceptMap = Json.decodeFromJsonElement<ConceptMap>(resource)
                    conceptMapService.registerConceptMap(conceptMap)
                    ProcessedCounts(conceptMaps = 1)
                } catch (e: Exception) {
                    ProcessedCounts(other = 1)
                }
            }
            "ValueSet" -> {
                try {
                    val valueSet = Json.decodeFromJsonElement<ValueSet>(resource)
                    valueSetService.registerValueSet(valueSet)
                    ProcessedCounts(valueSets = 1)
                } catch (e: Exception) {
                    ProcessedCounts(other = 1)
                }
            }
            "CodeSystem" -> {
                try {
                    val codeSystem = Json.decodeFromJsonElement<CodeSystem>(resource)
                    codeSystemService.registerCodeSystem(codeSystem)
                    ProcessedCounts(codeSystems = 1)
                } catch (e: Exception) {
                    ProcessedCounts(other = 1)
                }
            }
            "StructureDefinition" -> {
                try {
                    val structureDefinition = Json.decodeFromJsonElement<StructureDefinition>(resource)
                    validationService.registerStructureDefinition(structureDefinition)
                    ProcessedCounts(structureDefinitions = 1)
                } catch (e: Exception) {
                    ProcessedCounts(other = 1)
                }
            }
            "StructureMap" -> {
                // StructureMap processing would be handled by the main FmlRunner
                ProcessedCounts(structureMaps = 1)
            }
            else -> {
                ProcessedCounts(other = 1)
            }
        }
    }

    /**
     * Update processing statistics
     */
    private fun updateStats(bundle: Bundle, processed: ProcessedCounts) {
        val totalResources = (bundle.entry?.size ?: 0)
        
        stats = stats.copy(
            totalBundles = stats.totalBundles + 1,
            totalResources = stats.totalResources + totalResources,
            lastProcessed = kotlinx.datetime.Clock.System.now().toString()
        )

        // Update resource counts
        resourceCounts["ConceptMap"] = (resourceCounts["ConceptMap"] ?: 0) + processed.conceptMaps
        resourceCounts["ValueSet"] = (resourceCounts["ValueSet"] ?: 0) + processed.valueSets
        resourceCounts["CodeSystem"] = (resourceCounts["CodeSystem"] ?: 0) + processed.codeSystems
        resourceCounts["StructureDefinition"] = (resourceCounts["StructureDefinition"] ?: 0) + processed.structureDefinitions
        resourceCounts["StructureMap"] = (resourceCounts["StructureMap"] ?: 0) + processed.structureMaps
        resourceCounts["Other"] = (resourceCounts["Other"] ?: 0) + processed.other

        stats = stats.copy(resourceCounts = resourceCounts.toMap())
    }

    /**
     * Get Bundle processing statistics
     */
    fun getStats(): BundleStats {
        return stats
    }

    /**
     * Clear all processed resources and reset statistics
     */
    fun clear() {
        conceptMapService.clear()
        valueSetService.clear()
        codeSystemService.clear()
        validationService.clear()
        resourceCounts.clear()
        stats = BundleStats()
    }

    /**
     * Get count of processed resources by type
     */
    fun getResourceCount(resourceType: String): Int {
        return resourceCounts[resourceType] ?: 0
    }

    /**
     * Get total count of processed resources
     */
    fun getTotalResourceCount(): Int {
        return resourceCounts.values.sum()
    }
}